hal/src/thumbv7em/icm.rs

```diff
@@ -0,0 +1,1612 @@
+        matches!(self.isr().read().rbe().bits() & self.mask(), 1)
+    }
+
+    /// Region Digest Mismatches!( interrupt status
+    #[inline]
+    pub fn get_rdm_int(&self) -> bool {
+        matches!(self.isr().read().rdm().bits() & self.mask(), 1)
+    }
+
+    /// Region Hash Completed interrupt status
+    #[inline]
+    pub fn get_rhc_int(&self) -> bool {
+        matches!(self.isr().read().rhc().bits() & self.mask(), 1)
+    }
+
+    /// When reading the interrupt (ISR) register, it is cleared
+    ///
+    /// This might result in unexpected results for example
+    /// when sequentially trying to determine which interrupt triggered.
+    ///
+    /// This is an alternative, return all the data from the register
+    /// and parse later with the designated `get_[name]_int` functions.
+    #[inline]
+    pub fn get_interrupt_status(&mut self) -> IcmRegionInterrupt<I> {
+        let interrupt_vector = self.isr().read().bits();
+        IcmRegionInterrupt {
+            region: PhantomData,
+            interrupt_vector,
+        }
+    }
+}
+
+/// ICM Peripheral
+///
+/// Encapsulates the PAC which acts as a token
+/// and provides an interface to the ICM hardware
+pub struct Icm {
+    /// IcmToken providing hardware access
+    token: IcmToken,
+}
+
+impl Icm {
+    /// Create the interface for the ICM peripheral
+    ///
+    /// Don't forget to enable the `APB` bus for ICM
+    ///
+    /// `AHB` bus is on by default at reset
+    ///
+    /// Clock::v1
+    /// `mclk.apbcmask.modify(|_, w| w.icm_().set_bit());`
+    ///
+    /// Clock::v2
+    /// `tokens.apbs.icm.enable();`
+    #[inline]
+    pub fn new(icm: crate::pac::ICM) -> Self {
+        let token = IcmToken::new(icm);
+        Self { token }
+    }
+
+    /// Enable the ICM peripheral
+    #[inline]
+    pub fn enable(&mut self) {
+        self.token.ctrl().write(|w| w.enable().set_bit());
+    }
+
+    /// Get enabled status of the ICM peripheral
+    #[inline]
+    pub fn icm_status(&self) -> bool {
+        self.token.sr().read().enable().bit_is_set()
+    }
+
+    /// Disable the ICM peripheral
+    #[inline]
+    pub fn disable(&mut self) {
+        self.token.ctrl().write(|w| w.disable().set_bit());
+    }
+
+    /// Reset the ICM controller
+    ///
+    /// Does not seem to clear DSCR, HASH addr
+    ///
+    /// The only way to clear the `URAD` and `URAT` fields
+    /// is by resetting the ICM controller
+    #[inline]
+    pub fn swrst(&mut self) {
+        self.token.ctrl().write(|w| w.swrst().set_bit());
+    }
+
+    /// Destroy the ICM peripheral and return the underlying ICM register
+    #[inline]
+    pub fn destroy(self) -> crate::pac::ICM {
+        self.token.icm
+    }
+
+    // Region specifics
+
+    /// Enable region0
+    ///
+    /// Creates an [`IcmRegionToken`] which provides region specific
+    /// settings
+    #[inline]
+    pub fn enable_region0(&mut self) -> IcmRegionToken<Region0> {
+        IcmRegionToken::new()
+    }
```

Comment:
Maybe consider also adding a generic version as well, i.e. `fn enable_region<R: RegionNum>(...)`? It might be useful to let the type be inferred sometimes.

Reply:
Added :+1: